include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh

```diff
@@ -0,0 +1,359 @@
+namespace detail {
+
+/**
+ * @brief Struct to store ranks of bits at 256-bit intervals
+ */
+struct rank {
+  uint32_t abs_hi_;                    ///< Upper 32 bits of base
+  uint8_t abs_lo_;                     ///< Lower 8 bits of base
+  cuda::std::array<uint8_t, 3> rels_;  ///< Offsets for 64-bit sub-intervals
+
+  /**
+   * @brief Gets base rank of current 256-bit interval
+   *
+   * @return The base rank
+   */
+  __host__ __device__ constexpr uint64_t abs() const noexcept
+  {
+    return (static_cast<uint64_t>(abs_hi_) << 8) | abs_lo_;
+  }
+
+  /**
+   * @brief Sets base rank of current 256-bit interval
+   *
+   * @param abs Base rank
+   */
+  __host__ __device__ constexpr void set_abs(uint64_t abs) noexcept
+  {
+    abs_hi_ = static_cast<uint32_t>(abs >> 8);
+    abs_lo_ = static_cast<uint8_t>(abs);
+  }
+};
+
+/**
+ * @brief Bitvector class with rank and select index structures
+ *
+ * In addition to standard bitvector get/set operations, this class provides
+ * rank and select operation API. It maintains index structures to make both these
+ * new operations close to constant time.
+ * Bitvector construction happens on host, after which the structures are moved to device.
+ * All subsequent read-only operations access device structures only.
+ *
+ * @tparam Allocator Type of allocator used for device storage
+ */
+// TODO: have to use device_malloc_allocator for now otherwise the container cannot grow
+template <class Allocator = thrust::device_malloc_allocator<std::byte>>
+class dynamic_bitset {
+ public:
+  using size_type = std::size_t;  ///< size type to specify bit index
+  using slot_type = uint64_t;     ///< Slot type
+  /// Type of the allocator to (de)allocate words
+  using allocator_type = typename std::allocator_traits<Allocator>::rebind_alloc<slot_type>;
+
+  static constexpr size_type words_per_block = 4;  ///< Tradeoff between space efficiency and perf.
+  static constexpr size_type bits_per_word   = sizeof(slot_type) * CHAR_BIT;     ///< Bits in a word
+  static constexpr size_type bits_per_block  = words_per_block * bits_per_word;  ///< Trivial
+
+  /**
+   * @brief Constructs an empty bitvector
+   *
+   * @param allocator Allocator used for allocating device storage
+   */
+  constexpr dynamic_bitset(Allocator const& allocator = Allocator{});
+
+  /**
+   * @brief adds a new bit at the end
+   *
+   * Grows internal storage if needed
+   *
+   * @param bit Boolean value of new bit to be added
+   */
+  constexpr void append(bool bit) noexcept;
+
+  /**
+   * @brief Modifies a single bit
+   *
+   * @param index position of bit to be modified
+   * @param bit new value of bit
+   */
+  constexpr void set(size_type index, bool bit) noexcept;
+
+  /**
+   * @brief Sets last bit to specified value
+   *
+   * @param bit new value of last bit
+   */
+  constexpr void set_last(bool bit) noexcept;
+
+  /**
+   * @brief Builds indexes for rank and select
+   */
+  constexpr void build() noexcept;
```

Comment:
I'm thinking of adding a new data member called
```cuda
  bool is_built_; ///< Flag indicating whether the rank and select indices are built or not
```
We set the flag to `false` in ctor and all modifier functions and at the beginning of query/read functions, we can do something like:
```cuda
  if (not is_built_) {
    this->build();
    is_built_ = true;
  }
```
In this way,  this could be a private function and we can hide the build process from users.

Reply:
This makes sense. We need to do this check within every bulk query operation and ref(). I will take a stab at making the change.

Reply:
Unless it's needed by trie, let's just focus on bulk operations for now since such a flag in ref APIs would cause host/device synchronization issues that we tend to ignore for now.

Reply:
a629730842d08d41696b